tests/integration/conftest.py

```diff
@@ -44,17 +48,17 @@ async def deploy(ops_test: OpsTest):
         raise_on_blocked=False,
         timeout=600,
     )
-    await ops_test.model.deploy(NGINX_NAME, trust=True)
 
     async with ops_test.fast_forward():
         await ops_test.model.wait_for_idle(
-            apps=[NGINX_NAME, POSTGRES_NAME],
+            apps=[NGINX_NAME, POSTGRES_NAME, REDIS_NAME],
             status="active",
             raise_on_blocked=False,
             timeout=600,
         )
 
         await perform_superset_integrations(ops_test)
+        time.sleep(30)  # wait for application to start
```

Comment:
**question:** What are you waiting for here?

Is there another, more reliable way to determine application 'start'? API call, Juju status etc?

Reply:
In the `kafka-operator` charm, we do the following: 

- In `_on_start` (basically `pebble-ready`) we call a method [_on_update_status()](https://github.com/canonical/kafka-operator/blob/db3fe14a45a4f539deeafe626f1ad30a475a5255/src/charm.py#L307-L326) after starting the service
- There, we check for (`broker_active()`)[https://github.com/canonical/kafka-operator/blob/db3fe14a45a4f539deeafe626f1ad30a475a5255/src/charm.py], which has a [retry + backoff](https://github.com/canonical/kafka-operator/blob/db3fe14a45a4f539deeafe626f1ad30a475a5255/src/utils.py#L27-L34) using `tenacity` that pings the actual service
    - Like your example, this also takes time to set-up
- If that check passes, THEN we set (`ActiveStatus`)[https://github.com/canonical/kafka-operator/blob/db3fe14a45a4f539deeafe626f1ad30a475a5255/src/charm.py#L224]

Reply:
`_on_update_status()` is also the event-handler for `UpdateStatus`, which will run (by default) every 5m. 

Basically, if we're claiming that the charm is 'active', we should be constantly checking that the service is actually running. In native K8s world, this would be the equivalent to having a `LivenessProbe`. I think we can configure the `Live/ReadinessProbe`s that Juju uses with ROCKs, but that's a fairly new feature I believe and far out of the scope of this.

Reply:
Makes sense, I've been working on this and will open it as a new PR as it will be applicable to more than just the Redis relation.